enhancements/ingress/dynamic-config-manager.md

```diff
@@ -0,0 +1,469 @@
+7. Allow DCM to soak as tech preview for at least 1 OCP release.
+8. Possibly add new logic to DCM to exploit new HAProxy 2.y features.
+9. Remove the featuregate and mark DCM as GA.
+
+Steps 5 and 8 are stretch goals and may be done in later OpenShift releases.
+The other steps are hard requirements, and we intend to complete them in
+OpenShift 4.18.
+
+### Workflow Description
+
+Initially, enabling Dynamic Config Manager (DCM) will require enabling a
+featuregate.  Ultimately, DCM should be enabled by default. This configuration
+should be completely transparent to the end-user.  The only effect should be
+that OpenShift router uses less CPU and memory, balances traffic more evenly
+after endpoints updates, and tracks metrics more accurately after route and
+endpoints updates.
+
+#### Variation and form factor considerations [optional]
+
+DCM should function the same on standalone OpenShift, MicroShift, and
+HyperShift.
+
+### API Extensions
+
+DCM does not require any API extensions beyond the featuregate.  However, it
+does have an unsupported config override in the IngressController API in case a
+critical issue is discovered after DCM has been enabled by default.
+
+### Implementation Details/Notes/Constraints [optional]
+
+DCM was implemented in OpenShift 3.  This initial implementation requires
+pre-allocating both backends and servers in order to accommodate routes and
+endpoints that are created after HAProxy starts.
+
+Backends are pre-allocated based on *blueprint routes*, which are route objects
+that the cluster-admin must create in a designated namespace.  A blueprint route
+must specify the TLS termination type and set of annotations.  Backends and
+servers are pre-allocated when HAProxy starts.
+
+If DCM cannot configure something dynamically, it falls back to the old
+fork-and-reload procedure.  In particular, the following conditions require
+falling back to fork and reload:
+
+- More routes are created than the number of pre-allocated backends.
+- More endpoints are created for a route than the number of pre-allocated server slots for that specific route's backend.
+- A route is created that does not match the TLS termination type and annotations of any blueprint route.
+
+In OpenShift 3, configuring blueprint routes was left up to the cluster-admin to
+do.  As such, blueprint routes constitute a user-facing API.  Additionally, the
+cluster-admin could configure the number of server slots to pre-allocate for
+each backend.
+
+Note that pre-allocated server slots and backends take up memory.  In a cluster
+with many routes, pre-allocating multiple server slots for each backend can take
+a considerably large amount of memory.  See
+https://gist.github.com/frobware/2b527ce3f040797909eff482a4776e0b for an
+analysis of the potential memory impact for different choices of balancing
+algorithm, number of threads, number of backends, and number of pre-allocated
+server slots per backend.
```

Comment:
It would be nice to see a brief summary that shows how fast memory grows for a typical use case.

Reply:
Do you mean pulling in Andy's table, or do you mean gathering more data so we can plot some curves?

Reply:
Neither - I meant summarizing the data in the table.  You could keep the reference to the data set.